Marathon stops apps instead of restart

I've got another issue in addition to #2360. This is how the group looks like after deployment to `PUT /v2/groups`:

![image](https://cloud.githubusercontent.com/assets/89186/10363515/d0734660-6dac-11e5-8148-eeecd48217ec.png)

Hierarchy looks like this:

* `/app/db` — 3 apps in this group, `minimumHealthCapacity` is 0
* `/app/engine` — 6 apps in this group, `minimumHealthCapacity` is 0
* `/app/svc` — 2 apps in this group, `minimumHealthCapacity` is 1

Let's take a look at one engine.

Timeline:

1. `11:03:11.108` app is running.
2. `11:05:04.767` deploy triggered, single instance killed and DEPLOY STOPPED.
3. `11:08:11.154` app started again.

Here are the logs for one app:

```
October 8th 2015, 11:08:35.487	[INFO] [10/08/2015 10:08:11.543] [marathon-akka.actor.default-dispatcher-23] [akka://marathon/user/launchQueue/2/9-app_engine_execution] No tasks left to launch. Stop receiving offers for /app/engine/execution, 2015-10-08T10:05:04.724Z
October 8th 2015, 11:08:34.378	[2015-10-08 10:08:11,542] INFO Offer [20150930-060945-269165578-5050-12932-O444787]. Insufficient resources for [/app/engine/execution] (need cpus=0.1, mem=1024.0, disk=0.0, ports=(), available in offer:
October 8th 2015, 11:08:33.286	[2015-10-08 10:08:11,541] INFO Offer [20150930-060945-269165578-5050-12932-O444786]. Insufficient resources for [/app/engine/execution] (need cpus=0.1, mem=1024.0, disk=0.0, ports=(), available in offer:
October 8th 2015, 11:08:32.355	[2015-10-08 10:08:11,491] INFO Offer [20150930-060945-269165578-5050-12932-O444785]. Insufficient resources for [/app/engine/execution] (need cpus=0.1, mem=1024.0, disk=0.0, ports=(), available in offer:
October 8th 2015, 11:08:31.118	[2015-10-08 10:08:11,490] INFO Offer [20150930-060945-269165578-5050-12932-O444784]. Insufficient resources for [/app/engine/execution] (need cpus=0.1, mem=1024.0, disk=0.0, ports=(), available in offer:
October 8th 2015, 11:08:30.009	[2015-10-08 10:08:11,489] INFO Offer [20150930-060945-269165578-5050-12932-O444783]. Insufficient resources for [/app/engine/execution] (need cpus=0.1, mem=1024.0, disk=0.0, ports=(), available in offer:
October 8th 2015, 11:08:28.690	[2015-10-08 10:08:11,448] INFO Offer [20150930-060945-269165578-5050-12932-O444782]. Insufficient resources for [/app/engine/execution] (need cpus=0.1, mem=1024.0, disk=0.0, ports=(), available in offer:
October 8th 2015, 11:08:27.023	[2015-10-08 10:08:11,446] INFO Offer [20150930-060945-269165578-5050-12932-O444781]. Insufficient resources for [/app/engine/execution] (need cpus=0.1, mem=1024.0, disk=0.0, ports=(), available in offer:
October 8th 2015, 11:08:25.298	[2015-10-08 10:08:11,445] INFO Offer [20150930-060945-269165578-5050-12932-O444780]. Insufficient resources for [/app/engine/execution] (need cpus=0.1, mem=1024.0, disk=0.0, ports=(), available in offer:
October 8th 2015, 11:08:22.980	[2015-10-08 10:08:11,444] INFO Offer [20150930-060945-269165578-5050-12932-O444779]. Insufficient resources for [/app/engine/execution] (need cpus=0.1, mem=1024.0, disk=0.0, ports=(), available in offer:
October 8th 2015, 11:08:21.309	[2015-10-08 10:08:11,442] INFO Offer [20150930-060945-269165578-5050-12932-O444778]. Insufficient resources for [/app/engine/execution] (need cpus=0.1, mem=1024.0, disk=0.0, ports=(), available in offer:
October 8th 2015, 11:08:17.243	[2015-10-08 10:08:11,409] INFO Offer [20150930-060945-269165578-5050-12932-O444777]. Insufficient resources for [/app/engine/execution] (need cpus=0.1, mem=1024.0, disk=0.0, ports=(), available in offer:
October 8th 2015, 11:08:15.654	[2015-10-08 10:08:11,407] INFO Offer [20150930-060945-269165578-5050-12932-O444776]. Insufficient resources for [/app/engine/execution] (need cpus=0.1, mem=1024.0, disk=0.0, ports=(), available in offer:
October 8th 2015, 11:08:15.282	[2015-10-08 10:08:11,407] INFO Offer [20150930-060945-269165578-5050-12932-O444775]. Insufficient resources for [/app/engine/execution] (need cpus=0.1, mem=1024.0, disk=0.0, ports=(), available in offer:
October 8th 2015, 11:08:13.957	[2015-10-08 10:08:11,406] INFO Offer [20150930-060945-269165578-5050-12932-O444774]. Insufficient resources for [/app/engine/execution] (need cpus=0.1, mem=1024.0, disk=0.0, ports=(), available in offer:
October 8th 2015, 11:08:13.030	[2015-10-08 10:08:11,405] INFO Offer [20150930-060945-269165578-5050-12932-O444773]. Insufficient resources for [/app/engine/execution] (need cpus=0.1, mem=1024.0, disk=0.0, ports=(), available in offer:
October 8th 2015, 11:08:11.973	[2015-10-08 10:08:11,405] INFO Offer [20150930-060945-269165578-5050-12932-O444772]. Insufficient resources for [/app/engine/execution] (need cpus=0.1, mem=1024.0, disk=0.0, ports=(), available in offer:
October 8th 2015, 11:08:11.154	[2015-10-08 10:08:11,108] INFO Queueing 1 new tasks for /app/engine/execution (0 queued or running) (mesosphere.marathon.SchedulerActions:520)
October 8th 2015, 11:08:11.107	[2015-10-08 10:08:11,105] INFO Need to scale /app/engine/execution from 0 up to 1 instances (mesosphere.marathon.SchedulerActions:514)
October 8th 2015, 11:05:17.835	[2015-10-08 10:05:11,247] INFO Increasing delay. Task launch delay for [/app/engine/execution] changed from [0 milliseconds] to [1 seconds]. (mesosphere.marathon.core.launchqueue.impl.RateLimiter$:52)
October 8th 2015, 11:05:17.701	[INFO] [10/08/2015 10:05:09.264] [marathon-akka.actor.default-dispatcher-30] [akka://marathon/user/MarathonScheduler/$a/DeploymentManager/8ee9acf7-c784-46d9-9efd-a68997a3d16b/$f] All tasks for /app/engine/execution are healthy
October 8th 2015, 11:05:17.607	[INFO] [10/08/2015 10:05:07.151] [marathon-akka.actor.default-dispatcher-54] [akka://marathon/user/launchQueue/2/9-app_engine_execution] No tasks left to launch. Stop receiving offers for /app/engine/execution, 2015-10-08T10:05:04.724Z
October 8th 2015, 11:05:17.509	[INFO] [10/08/2015 10:05:06.167] [marathon-akka.actor.default-dispatcher-43] [akka://marathon/user/MarathonScheduler/$a/DeploymentManager/8ee9acf7-c784-46d9-9efd-a68997a3d16b/$f] Reconciliating tasks during app /app/engine/execution restart: queuing 1 new tasks
October 8th 2015, 11:05:17.509	[ERROR] [10/08/2015 10:05:06.167] [marathon-akka.actor.default-dispatcher-43] [akka://marathon/user/MarathonScheduler/$a/DeploymentManager/8ee9acf7-c784-46d9-9efd-a68997a3d16b/$f] Old task app_engine_execution.9b6ad4d5-6da1-11e5-9e56-000f532c02a0 failed on slave 20150908-164039-319497226-5050-29034-S5 during app /app/engine/execution restart
October 8th 2015, 11:05:17.509	[INFO] [10/08/2015 10:05:06.167] [marathon-akka.actor.default-dispatcher-52] [akka://marathon/user/launchQueue/2/9-app_engine_execution] scaling change for '/app/engine/execution', version 2015-10-08T10:05:04.724Z with 1 initial tasks
October 8th 2015, 11:05:05.009	[WARN] [10/08/2015 10:05:04.811] [marathon-akka.actor.default-dispatcher-57] [akka://marathon/user/launchQueue/2/9-app_engine_execution] Received delay update for other app: DelayUpdate(AppDefinition(/app/engine/execution,None,None,None,Map(REDACTED -> true),1,0.1,1024.0,0.0,,Set(field: "hostname"
October 8th 2015, 11:05:04.963	[INFO] [10/08/2015 10:05:04.810] [marathon-akka.actor.default-dispatcher-59] [akka://marathon/user/MarathonScheduler/$a/DeploymentManager/8ee9acf7-c784-46d9-9efd-a68997a3d16b/$f] For minimumHealthCapacity 0.0 of /app/engine/execution leave 0 tasks running, maximum capacity 1, killing 1 tasks immediately
October 8th 2015, 11:05:04.921	* Restart(App(/app/engine/execution, image="mydocker.registry/myapp/execution:1.2.3.4.5")))
October 8th 2015, 11:05:04.875	[2015-10-08 10:05:04,757] INFO [/app/engine/execution] storing new app version 2015-10-08T10:05:04.724Z (mesosphere.marathon.state.GroupManager:144)
October 8th 2015, 11:05:04.874	* Restart(App(/app/engine/execution, image="mydocker.registry/myapp/execution:1.2.3.4.5")))
October 8th 2015, 11:05:04.831	* Restart(App(/app/engine/execution, image="mydocker.registry/myapp/execution:1.2.3.4.5")))
October 8th 2015, 11:05:04.767	[2015-10-08 10:05:04,730] INFO [/app/engine/execution]: restart detected for app (oldVersion FullVersionInfo(2015-10-08T09:47:36.555Z,2015-10-08T09:47:36.555Z,2015-09-25T16:02:41.400Z)) (mesosphere.marathon.upgrade.GroupVersioningUtil$:38)
October 8th 2015, 11:03:11.108	[2015-10-08 10:03:11,106] INFO Already running 1 instances of /app/engine/execution. Not scaling. (mesosphere.marathon.SchedulerActions:538)
October 8th 2015, 10:58:11.150	[2015-10-08 09:58:11,106] INFO Already running 1 instances of /app/engine/execution. Not scaling. (mesosphere.marathon.SchedulerActions:538)
October 8th 2015, 10:53:11.138	[2015-10-08 09:53:11,109] INFO Already running 1 instances of /app/engine/execution. Not scaling. (mesosphere.marathon.SchedulerActions:538)
```

Marathon 0.11.0, Mesos 0.23.0. I've seen similar on 0.10.0, but can't be sure.

Tried to replicate locally with no luck.

--- a/src/main/scala/mesosphere/marathon/MarathonScheduler.scala
+++ b/src/main/scala/mesosphere/marathon/MarathonScheduler.scala
@@ -76,11 +76,7 @@ class MarathonScheduler @Inject() (
     log.info("Received status update for task %s: %s (%s)"
       .format(status.getTaskId.getValue, status.getState, status.getMessage))
 
-    val update: TaskStatusUpdate = TaskStatusUpdate(
-      timestamp = clock.now(), status.getTaskId,
-      MarathonTaskStatus(status)
-    )
-    taskStatusProcessor.publish(update).onFailure {
+    taskStatusProcessor.publish(status).onFailure {
       case NonFatal(e) =>
         log.error(s"while processing task status update $status", e)
     }

--- a/src/main/scala/mesosphere/marathon/core/CoreGuiceModule.scala
+++ b/src/main/scala/mesosphere/marathon/core/CoreGuiceModule.scala
@@ -7,8 +7,19 @@ import mesosphere.marathon.core.launcher.OfferProcessor
 import mesosphere.marathon.core.launchqueue.LaunchQueue
 import mesosphere.marathon.core.leadership.{ LeadershipCoordinator, LeadershipModule }
 import mesosphere.marathon.core.task.bus.{ TaskStatusEmitter, TaskStatusObservables }
+import mesosphere.marathon.core.task.tracker.impl.steps.{
+  TaskStatusEmitterPublishStepImpl,
+  UpdateTaskTrackerStepImpl,
+  ScaleAppUpdateStepImpl,
+  PostToEventStreamStepImpl,
+  NotifyRateLimiterStepImpl,
+  NotifyLaunchQueueStepImpl,
+  NotifyHealthCheckManagerStepImpl,
+  ContinueOnErrorStep,
+  AcknowledgeTaskUpdateStepImpl
+}
 import mesosphere.marathon.core.task.tracker.impl.TaskStatusUpdateProcessorImpl
-import mesosphere.marathon.core.task.tracker.{ TaskStatusUpdateProcessor, TaskTrackerModule }
+import mesosphere.marathon.core.task.tracker.{ TaskStatusUpdateProcessor, TaskStatusUpdateStep, TaskTrackerModule }
 
 /**
   * Provides the glue between guice and the core modules.
@@ -44,6 +55,34 @@ class CoreGuiceModule extends AbstractModule {
 
   @Provides @Singleton
   final def appInfoService(appInfoModule: AppInfoModule): AppInfoService = appInfoModule.appInfoService
+
+  @Provides @Singleton
+  def taskStatusUpdateSteps(
+    notifyHealthCheckManagerStepImpl: NotifyHealthCheckManagerStepImpl,
+    acknowledgeTaskUpdateStepImpl: AcknowledgeTaskUpdateStepImpl,
+    notifyRateLimiterStepImpl: NotifyRateLimiterStepImpl,
+    taskStatusEmitterPublishImpl: TaskStatusEmitterPublishStepImpl,
+    postToEventStreamStepImpl: PostToEventStreamStepImpl,
+    scaleAppUpdateStepImpl: ScaleAppUpdateStepImpl,
+    notifyLaunchQueueStepImpl: NotifyLaunchQueueStepImpl,
+    updateTaskTrackerStepImpl: UpdateTaskTrackerStepImpl): Seq[TaskStatusUpdateStep] = {
+
+    // This is a sequence on purpose. The specified steps are executed in order for every
+    // task status update.
+    // This way we make sure that e.g. the taskTracker already reflects the changes for the update
+    // (updateTaskTrackerStepImpl) before we notify the launch queue (notifyLaunchQueueStepImpl).
+
+    Seq(
+      ContinueOnErrorStep(notifyHealthCheckManagerStepImpl),
+      ContinueOnErrorStep(notifyRateLimiterStepImpl),
+      updateTaskTrackerStepImpl,
+      ContinueOnErrorStep(notifyLaunchQueueStepImpl),
+      ContinueOnErrorStep(taskStatusEmitterPublishImpl),
+      ContinueOnErrorStep(postToEventStreamStepImpl),
+      ContinueOnErrorStep(scaleAppUpdateStepImpl),
+      acknowledgeTaskUpdateStepImpl
+    )
+  }
 
   override def configure(): Unit = {
     bind(classOf[Clock]).toInstance(Clock())

--- a/src/main/scala/mesosphere/marathon/core/CoreModuleImpl.scala
+++ b/src/main/scala/mesosphere/marathon/core/CoreModuleImpl.scala
@@ -74,7 +74,6 @@ class CoreModuleImpl @Inject() (
 
     // internal core dependencies
     offerMatcherManagerModule.subOfferMatcherManager,
-    taskBusModule.taskStatusObservables,
     maybeOfferReviver,
 
     // external guice dependencies

--- a/src/main/scala/mesosphere/marathon/core/launchqueue/LaunchQueue.scala
+++ b/src/main/scala/mesosphere/marathon/core/launchqueue/LaunchQueue.scala
@@ -1,9 +1,11 @@
 package mesosphere.marathon.core.launchqueue
 
 import mesosphere.marathon.core.launchqueue.LaunchQueue.QueuedTaskCount
+import mesosphere.marathon.core.task.bus.TaskStatusObservables.TaskStatusUpdate
 import mesosphere.marathon.state.{ AppDefinition, PathId, Timestamp }
 
 import scala.collection.immutable.Seq
+import scala.concurrent.Future
 
 object LaunchQueue {
 
@@ -53,4 +55,7 @@ trait LaunchQueue {
 
   /** Reset the backoff delay for the given AppDefinition. */
   def resetDelay(app: AppDefinition): Unit
+
+  /** Notify queue about TaskUpdate */
+  def notifyOfTaskUpdate(update: TaskStatusUpdate): Future[Option[QueuedTaskCount]]
 }

--- a/src/main/scala/mesosphere/marathon/core/launchqueue/LaunchQueueModule.scala
+++ b/src/main/scala/mesosphere/marathon/core/launchqueue/LaunchQueueModule.scala
@@ -4,15 +4,14 @@ import akka.actor.{ ActorRef, Props }
 import mesosphere.marathon.core.base.Clock
 import mesosphere.marathon.core.flow.OfferReviver
 import mesosphere.marathon.core.launchqueue.impl.{
-  LaunchQueueDelegate,
   AppTaskLauncherActor,
   LaunchQueueActor,
+  LaunchQueueDelegate,
   RateLimiter,
   RateLimiterActor
 }
 import mesosphere.marathon.core.leadership.LeadershipModule
 import mesosphere.marathon.core.matcher.manager.OfferMatcherManager
-import mesosphere.marathon.core.task.bus.TaskStatusObservables
 import mesosphere.marathon.state.{ AppDefinition, AppRepository }
 import mesosphere.marathon.tasks.{ TaskFactory, TaskTracker }
 
@@ -24,7 +23,6 @@ class LaunchQueueModule(
     leadershipModule: LeadershipModule,
     clock: Clock,
     subOfferMatcherManager: OfferMatcherManager,
-    taskStatusObservables: TaskStatusObservables,
     maybeOfferReviver: Option[OfferReviver],
     appRepository: AppRepository,
     taskTracker: TaskTracker,
@@ -50,7 +48,6 @@ class LaunchQueueModule(
       subOfferMatcherManager,
       clock,
       taskFactory,
-      taskStatusObservables,
       maybeOfferReviver,
       taskTracker,
       rateLimiterActor)(app, count)

--- a/src/main/scala/mesosphere/marathon/core/launchqueue/impl/AppTaskLauncherActor.scala
+++ b/src/main/scala/mesosphere/marathon/core/launchqueue/impl/AppTaskLauncherActor.scala
@@ -31,7 +31,6 @@ private[launchqueue] object AppTaskLauncherActor {
     offerMatcherManager: OfferMatcherManager,
     clock: Clock,
     taskFactory: TaskFactory,
-    taskStatusObservable: TaskStatusObservables,
     maybeOfferReviver: Option[OfferReviver],
     taskTracker: TaskTracker,
     rateLimiterActor: ActorRef)(
@@ -40,7 +39,7 @@ private[launchqueue] object AppTaskLauncherActor {
     Props(new AppTaskLauncherActor(
       config,
       offerMatcherManager,
-      clock, taskFactory, taskStatusObservable,
+      clock, taskFactory,
       maybeOfferReviver,
       taskTracker, rateLimiterActor,
       app, initialCount))
@@ -83,7 +82,6 @@ private class AppTaskLauncherActor(
     offerMatcherManager: OfferMatcherManager,
     clock: Clock,
     taskFactory: TaskFactory,
-    taskStatusObservable: TaskStatusObservables,
     maybeOfferReviver: Option[OfferReviver],
     taskTracker: TaskTracker,
     rateLimiterActor: ActorRef,
@@ -96,8 +94,6 @@ private class AppTaskLauncherActor(
   private[this] var recheckBackOff: Option[Cancellable] = None
   private[this] var backOffUntil: Option[Timestamp] = None
 
-  /** Receive task updates to keep task list up-to-date. */
-  private[this] var taskStatusUpdateSubscription: Subscription = _
   /** Currently known running tasks or tasks that we have requested to be launched. */
   private[this] var runningTasks: Set[MarathonTask] = _
   /** Like runningTasks but indexed by the taskId String */
@@ -112,10 +108,6 @@ private class AppTaskLauncherActor(
     log.info("Started appTaskLaunchActor for {} version {} with initial count {}",
       app.id, app.version, tasksToLaunch)
 
-    taskStatusUpdateSubscription = taskStatusObservable.forAppId(app.id).subscribe { update =>
-      log.debug("update {}", update)
-      self ! update
-    }
     runningTasks = taskTracker.get(app.id)
     runningTasksMap = runningTasks.map(task => task.getId -> task).toMap
 
@@ -123,7 +115,6 @@ private class AppTaskLauncherActor(
   }
 
   override def postStop(): Unit = {
-    taskStatusUpdateSubscription.unsubscribe()
     OfferMatcherRegistration.unregister()
     recheckBackOff.foreach(_.cancel())
 
@@ -247,6 +238,8 @@ private class AppTaskLauncherActor(
         maybeOfferReviver.foreach(_.reviveOffers())
       }
 
+      replyWithQueuedTaskCount()
+
     case TaskStatusUpdate(_, taskId, status) =>
       runningTasksMap.get(taskId.getValue) match {
         case None =>
@@ -263,6 +256,7 @@ private class AppTaskLauncherActor(
           runningTasksMap += taskId.getValue -> updatedTask
       }
 
+      replyWithQueuedTaskCount()
   }
 
   private[this] def removeTask(taskId: TaskID): Unit = {

--- a/src/main/scala/mesosphere/marathon/core/launchqueue/impl/LaunchQueueActor.scala
+++ b/src/main/scala/mesosphere/marathon/core/launchqueue/impl/LaunchQueueActor.scala
@@ -15,6 +15,7 @@ import akka.event.LoggingReceive
 import akka.pattern.{ ask, pipe }
 import akka.util.Timeout
 import mesosphere.marathon.core.launchqueue.{ LaunchQueueConfig, LaunchQueue }
+import mesosphere.marathon.core.task.bus.TaskStatusObservables.TaskStatusUpdate
 import mesosphere.marathon.state.{ AppDefinition, PathId }
 import LaunchQueue.QueuedTaskCount
 
@@ -62,7 +63,9 @@ private[impl] class LaunchQueueActor(
   override def receive: Receive = LoggingReceive {
     Seq(
       receiveHandlePurging,
+      receiveTaskUpdateToSuspendedActor,
       receiveMessagesToSuspendedActor,
+      receiveTaskUpdate,
       receiveHandleNormalCommands
     ).reduce(_.orElse[Any, Unit](_))
   }
@@ -112,6 +115,11 @@ private[impl] class LaunchQueueActor(
       }
   }
 
+  private[this] def receiveTaskUpdateToSuspendedActor: Receive = {
+    case update: TaskStatusUpdate if suspendedLauncherPathIds(update.appId) =>
+      deferMessageToSuspendedActor(update, update.appId)
+  }
+
   private[this] def receiveMessagesToSuspendedActor: Receive = {
     case msg @ Count(appId) if suspendedLauncherPathIds(appId) =>
       deferMessageToSuspendedActor(msg, appId)
@@ -128,6 +136,18 @@ private[impl] class LaunchQueueActor(
     val deferredMessages: Vector[DeferredMessage] =
       suspendedLaunchersMessages(actorRef) :+ DeferredMessage(sender(), msg)
     suspendedLaunchersMessages += actorRef -> deferredMessages
+  }
+
+  private[this] def receiveTaskUpdate: Receive = {
+    case update: TaskStatusUpdate =>
+      import context.dispatcher
+      launchers.get(update.appId) match {
+        case Some(actorRef) =>
+          val eventualCount: Future[QueuedTaskCount] =
+            (actorRef ? update).mapTo[QueuedTaskCount]
+          eventualCount.map(Some(_)).pipeTo(sender())
+        case None => sender() ! None
+      }
   }
 
   private[this] def receiveHandleNormalCommands: Receive = {

--- a/src/main/scala/mesosphere/marathon/core/launchqueue/impl/LaunchQueueDelegate.scala
+++ b/src/main/scala/mesosphere/marathon/core/launchqueue/impl/LaunchQueueDelegate.scala
@@ -3,13 +3,14 @@ package mesosphere.marathon.core.launchqueue.impl
 import akka.actor.ActorRef
 import akka.pattern.ask
 import akka.util.Timeout
-import mesosphere.marathon.core.launchqueue.{ LaunchQueueConfig, LaunchQueue }
+import mesosphere.marathon.core.launchqueue.LaunchQueue.QueuedTaskCount
+import mesosphere.marathon.core.launchqueue.{ LaunchQueue, LaunchQueueConfig }
+import mesosphere.marathon.core.task.bus.TaskStatusObservables.TaskStatusUpdate
 import mesosphere.marathon.state.{ AppDefinition, PathId }
-import LaunchQueue.QueuedTaskCount
 
 import scala.collection.immutable.Seq
-import scala.concurrent.Await
-import scala.concurrent.duration.{ Deadline, _ }
+import scala.concurrent.duration._
+import scala.concurrent.{ Await, Future }
 import scala.util.control.NonFatal
 
 private[launchqueue] class LaunchQueueDelegate(
@@ -25,6 +26,9 @@ private[launchqueue] class LaunchQueueDelegate(
   override def get(appId: PathId): Option[QueuedTaskCount] =
     askQueueActor("get")(LaunchQueueDelegate.Count(appId)).asInstanceOf[Option[QueuedTaskCount]]
 
+  override def notifyOfTaskUpdate(update: TaskStatusUpdate): Future[Option[QueuedTaskCount]] =
+    askQueueActorFuture("notifyOfTaskUpdate")(update).mapTo[Option[QueuedTaskCount]]
+
   override def count(appId: PathId): Int = get(appId).map(_.tasksLeftToLaunch).getOrElse(0)
 
   override def listApps: Seq[AppDefinition] = list.map(_.app)
@@ -33,13 +37,18 @@ private[launchqueue] class LaunchQueueDelegate(
   override def add(app: AppDefinition, count: Int): Unit = askQueueActor("add")(LaunchQueueDelegate.Add(app, count))
 
   private[this] def askQueueActor[T](method: String)(message: T): Any = {
+    val answerFuture: Future[Any] = askQueueActorFuture(method)(message)
+    Await.result(answerFuture, config.launchQueueRequestTimeout().milliseconds)
+  }
+
+  private[this] def askQueueActorFuture[T](method: String)(message: T): Future[Any] = {
     implicit val timeout: Timeout = 1.second
     val answerFuture = actorRef ? message
     import scala.concurrent.ExecutionContext.Implicits.global
     answerFuture.recover {
       case NonFatal(e) => throw new RuntimeException(s"in $method", e)
     }
-    Await.result(answerFuture, config.launchQueueRequestTimeout().milliseconds)
+    answerFuture
   }
 
   override def addDelay(app: AppDefinition): Unit = rateLimiterRef ! RateLimiterActor.AddDelay(app)

--- a/src/main/scala/mesosphere/marathon/core/task/tracker/TaskStatusUpdateProcessor.scala
+++ b/src/main/scala/mesosphere/marathon/core/task/tracker/TaskStatusUpdateProcessor.scala
@@ -1,9 +1,9 @@
 package mesosphere.marathon.core.task.tracker
 
-import mesosphere.marathon.core.task.bus.TaskStatusObservables
+import org.apache.mesos.Protos.TaskStatus
 
 import scala.concurrent.Future
 
 trait TaskStatusUpdateProcessor {
-  def publish(status: TaskStatusObservables.TaskStatusUpdate): Future[Unit]
+  def publish(status: TaskStatus): Future[Unit]
 }

--- a/src/main/scala/mesosphere/marathon/core/task/tracker/TaskStatusUpdateStep.scala
+++ b/src/main/scala/mesosphere/marathon/core/task/tracker/TaskStatusUpdateStep.scala
@@ -1,0 +1,19 @@
+package mesosphere.marathon.core.task.tracker
+
+import mesosphere.marathon.Protos.MarathonTask
+import mesosphere.marathon.state.{ PathId, Timestamp }
+import org.apache.mesos.Protos.TaskStatus
+
+import scala.concurrent.Future
+
+/**
+  * A consumer interested in processing task status updates.
+  *
+  * There is a list of these which is called in sequence by the TaskStatusUpdateProcessor for every update.
+  */
+trait TaskStatusUpdateStep {
+  def name: String
+
+  def processUpdate(
+    timestamp: Timestamp, appId: PathId, maybeTask: Option[MarathonTask], mesosStatus: TaskStatus): Future[_]
+}

--- a/src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskStatusUpdateProcessorImpl.scala
+++ b/src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskStatusUpdateProcessorImpl.scala
@@ -1,133 +1,73 @@
 package mesosphere.marathon.core.task.tracker.impl
 
-import javax.inject.Named
+import javax.inject.Inject
 
-import akka.actor.ActorRef
-import akka.event.EventStream
-import com.google.inject.Inject
-import mesosphere.marathon.MarathonSchedulerActor.ScaleApp
-import mesosphere.marathon.MarathonSchedulerDriverHolder
 import mesosphere.marathon.Protos.MarathonTask
-import mesosphere.marathon.core.launchqueue.LaunchQueue
-import mesosphere.marathon.core.task.bus.MarathonTaskStatus.WithMesosStatus
-import mesosphere.marathon.core.task.bus.TaskStatusObservables.TaskStatusUpdate
-import mesosphere.marathon.core.task.bus.{ TaskStatusEmitter, TaskStatusObservables }
-import mesosphere.marathon.core.task.tracker.TaskStatusUpdateProcessor
-import mesosphere.marathon.event.{ EventModule, MesosStatusUpdateEvent }
-import mesosphere.marathon.health.HealthCheckManager
-import mesosphere.marathon.state.{ PathId, AppRepository, Timestamp }
+import mesosphere.marathon.core.base.Clock
+import mesosphere.marathon.core.task.tracker.{ TaskStatusUpdateProcessor, TaskStatusUpdateStep }
+import mesosphere.marathon.metrics.Metrics.Timer
+import mesosphere.marathon.metrics.{ MetricPrefixes, Metrics }
+import mesosphere.marathon.state.{ PathId, Timestamp }
 import mesosphere.marathon.tasks.{ TaskIdUtil, TaskTracker }
-import org.apache.mesos.Protos.{ TaskState, TaskStatus }
+import org.apache.mesos.Protos.TaskStatus
 import org.slf4j.LoggerFactory
 
 import scala.concurrent.Future
-import scala.util.control.NonFatal
 
 /**
-  * Processes task status update events, mostly to update the task tracker.
+  * Executes the given TaskStatusUpdateSteps for every update.
   */
 class TaskStatusUpdateProcessorImpl @Inject() (
-    taskStatusEmitter: TaskStatusEmitter,
-    appRepository: AppRepository,
-    launchQueue: LaunchQueue,
-    @Named(EventModule.busName) eventBus: EventStream,
-    @Named("schedulerActor") schedulerActor: ActorRef,
+    metrics: Metrics,
+    clock: Clock,
     taskIdUtil: TaskIdUtil,
-    healthCheckManager: HealthCheckManager,
     taskTracker: TaskTracker,
-    marathonSchedulerDriverHolder: MarathonSchedulerDriverHolder) extends TaskStatusUpdateProcessor {
+    steps: Seq[TaskStatusUpdateStep]) extends TaskStatusUpdateProcessor {
+  import scala.concurrent.ExecutionContext.Implicits.global
 
   private[this] val log = LoggerFactory.getLogger(getClass)
 
-  //TODO: fix style issue and enable this scalastyle check
-  //scalastyle:off cyclomatic.complexity method.length
-  def publish(incoming: TaskStatusObservables.TaskStatusUpdate): Future[Unit] = incoming match {
-    case TaskStatusUpdate(timestamp, taskId, WithMesosStatus(status)) =>
-      val appId = taskIdUtil.appId(taskId)
+  private[this] val publishFutureTimer: Timer =
+    metrics.timer(metrics.name(MetricPrefixes.SERVICE, getClass, "publishFuture"))
 
-      // forward health changes to the health check manager
-      val maybeTask = taskTracker.fetchTask(appId, taskId.getValue)
-      for (marathonTask <- maybeTask)
-        healthCheckManager.update(status, Timestamp(marathonTask.getVersion))
+  private[this] val stepTimers: Map[String, Timer] = steps.map { step =>
+    step.name -> metrics.timer(metrics.name(MetricPrefixes.SERVICE, getClass, s"step-${step.name}"))
+  }.toMap
 
-      import org.apache.mesos.Protos.TaskState._
+  log.info("Started status update processor with steps:\n{}", steps.map(step => s"* ${step.name}").mkString("\n"))
 
-      import scala.concurrent.ExecutionContext.Implicits.global
-      val eventuallyTracked = status.getState match {
-        case TASK_ERROR | TASK_FAILED | TASK_FINISHED | TASK_KILLED | TASK_LOST =>
-          // Remove from our internal list
-          taskTracker.terminated(appId, taskId.getValue).flatMap { taskOption =>
-            rateLimit(appId, status, taskOption).map { _ =>
-              schedulerActor ! ScaleApp(appId)
+  override def publish(status: TaskStatus): Future[Unit] = publishFutureTimer.timeFuture {
+    val now = clock.now()
+    val taskId = status.getTaskId
+    val appId = taskIdUtil.appId(taskId)
 
-              taskOption match {
-                case Some(task) => postEvent(status, task)
-                case None       => log.warn(s"Task not found. Do not post event for '{}'", taskId.getValue)
-              }
-            }
-          }
+    val maybeTask = taskTracker.fetchTask(appId, taskId.getValue)
 
-        case TASK_RUNNING if !maybeTask.exists(_.hasStartedAt) => // staged, not running
-          taskTracker.running(appId, status)
-            .map { task => postEvent(status, task) }
-            .recover {
-              case NonFatal(t) =>
-                log.warn(s"Task could not be saved. Do not post event for '${taskId.getValue}'", t)
-                driverOpt.foreach(_.killTask(status.getTaskId))
-            }
-
-        case TASK_STAGING if !taskTracker.contains(appId) =>
-          log.warn(s"Received status update for unknown app $appId, killing task ${status.getTaskId}")
-          driverOpt.foreach(_.killTask(status.getTaskId))
-          Future.successful(())
-
-        case _ =>
-          taskTracker.statusUpdate(appId, status).map {
-            case None =>
-              log.warn(s"Killing task ${status.getTaskId}")
-              driverOpt.foreach(_.killTask(status.getTaskId))
-            case _ =>
-          }
-      }
-      eventuallyTracked.map { _ =>
-        taskStatusEmitter.publish(incoming)
-        driverOpt.foreach(_.acknowledgeStatusUpdate(status))
-      }
-  }
-
-  private[this] def rateLimit(appId: PathId, status: TaskStatus, taskOption: Option[MarathonTask]): Future[Unit] = {
-    taskOption match {
-      case Some(task) if status.getState != TaskState.TASK_KILLED =>
-        import scala.concurrent.ExecutionContext.Implicits.global
-        appRepository.app(appId, Timestamp(task.getVersion)).map { appOption =>
-          // It would be nice if we could make sure that the delay gets send
-          // to the AppTaskLauncherActor before we continue but that would require quite some work.
-          //
-          // In production, it should not matter if we do not delay some task launches so much.
-          appOption.foreach(launchQueue.addDelay)
-        }
-      case _ =>
-        Future.successful(())
-    }
-  }
-
-  private[this] def driverOpt = marathonSchedulerDriverHolder.driver
-
-  private[this] def postEvent(status: TaskStatus, task: MarathonTask): Unit = {
-    log.info("Sending event notification.")
-    import scala.collection.JavaConverters._
-    eventBus.publish(
-      MesosStatusUpdateEvent(
-        status.getSlaveId.getValue,
-        status.getTaskId.getValue,
-        status.getState.name,
-        if (status.hasMessage) status.getMessage else "",
-        taskIdUtil.appId(task.getId),
-        task.getHost,
-        task.getPortsList.asScala,
-        task.getVersion
-      )
+    processUpdate(
+      timestamp = now,
+      appId = appId,
+      maybeTask = maybeTask,
+      mesosStatus = status
     )
   }
 
+  private[this] def processUpdate(
+    timestamp: Timestamp,
+    appId: PathId,
+    maybeTask: Option[MarathonTask],
+    mesosStatus: TaskStatus): Future[Unit] = {
+    steps.foldLeft(Future.successful(())) { (resultSoFar, nextStep) =>
+      stepTimers(nextStep.name).timeFuture {
+        resultSoFar.flatMap { _ =>
+          log.debug("Executing {} for [{}]", Array[Object](nextStep.name, mesosStatus.getTaskId.getValue): _*)
+          nextStep.processUpdate(timestamp, appId, maybeTask, mesosStatus).map { _ =>
+            log.debug(
+              "Done with executing {} for [{}]",
+              Array[Object](nextStep.name, mesosStatus.getTaskId.getValue): _*
+            )
+          }
+        }
+      }
+    }
+  }
 }

--- a/src/main/scala/mesosphere/marathon/core/task/tracker/impl/steps/AcknowledgeTaskUpdateStepImpl.scala
+++ b/src/main/scala/mesosphere/marathon/core/task/tracker/impl/steps/AcknowledgeTaskUpdateStepImpl.scala
@@ -1,0 +1,27 @@
+package mesosphere.marathon.core.task.tracker.impl.steps
+
+import javax.inject.Inject
+
+import mesosphere.marathon.MarathonSchedulerDriverHolder
+import mesosphere.marathon.Protos.MarathonTask
+import mesosphere.marathon.core.task.tracker.TaskStatusUpdateStep
+import mesosphere.marathon.state.{ PathId, Timestamp }
+import org.apache.mesos.Protos.TaskStatus
+
+import scala.concurrent.Future
+
+/**
+  * Send acknowledgement for this update to Mesos.
+  */
+class AcknowledgeTaskUpdateStepImpl @Inject() (marathonSchedulerDriverHolder: MarathonSchedulerDriverHolder)
+    extends TaskStatusUpdateStep {
+  override def name: String = "acknowledgeTaskUpdate"
+
+  override def processUpdate(
+    timestamp: Timestamp, appId: PathId, maybeTask: Option[MarathonTask], status: TaskStatus): Future[_] = {
+    driverOpt.foreach(_.acknowledgeStatusUpdate(status))
+    Future.successful(())
+  }
+
+  private[this] def driverOpt = marathonSchedulerDriverHolder.driver
+}

--- a/src/main/scala/mesosphere/marathon/core/task/tracker/impl/steps/ContinueOnErrorStep.scala
+++ b/src/main/scala/mesosphere/marathon/core/task/tracker/impl/steps/ContinueOnErrorStep.scala
@@ -1,0 +1,34 @@
+package mesosphere.marathon.core.task.tracker.impl.steps
+
+import mesosphere.marathon.Protos.MarathonTask
+import mesosphere.marathon.core.task.tracker.TaskStatusUpdateStep
+import mesosphere.marathon.state.{ AppDefinition, PathId, Timestamp }
+import org.apache.mesos.Protos.TaskStatus
+import org.slf4j.LoggerFactory
+
+import scala.concurrent.Future
+import scala.util.control.NonFatal
+
+/**
+  * Log errors in the wrapped step but do not fail because of them.
+  */
+class ContinueOnErrorStep(wrapped: TaskStatusUpdateStep) extends TaskStatusUpdateStep {
+  private[this] val log = LoggerFactory.getLogger(getClass)
+
+  override def name: String = s"continueOnError(${wrapped.name}})"
+
+  override def processUpdate(
+    timestamp: Timestamp, appId: PathId, maybeTask: Option[MarathonTask], mesosStatus: TaskStatus): Future[_] = {
+
+    import scala.concurrent.ExecutionContext.Implicits.global
+    wrapped.processUpdate(timestamp, appId, maybeTask, mesosStatus).recover {
+      case NonFatal(e) =>
+        log.error("while executing step {} for [{}], continue with other steps",
+          wrapped.name, mesosStatus.getTaskId.getValue, e)
+    }
+  }
+}
+
+object ContinueOnErrorStep {
+  def apply(wrapped: TaskStatusUpdateStep): ContinueOnErrorStep = new ContinueOnErrorStep(wrapped)
+}

--- a/src/main/scala/mesosphere/marathon/core/task/tracker/impl/steps/NotifyHealthCheckManagerStepImpl.scala
+++ b/src/main/scala/mesosphere/marathon/core/task/tracker/impl/steps/NotifyHealthCheckManagerStepImpl.scala
@@ -1,0 +1,27 @@
+package mesosphere.marathon.core.task.tracker.impl.steps
+
+import com.google.inject.Inject
+import mesosphere.marathon.Protos.MarathonTask
+import mesosphere.marathon.core.task.tracker.TaskStatusUpdateStep
+import mesosphere.marathon.health.HealthCheckManager
+import mesosphere.marathon.state.{ PathId, Timestamp }
+import org.apache.mesos.Protos.TaskStatus
+
+import scala.concurrent.Future
+
+/**
+  * Notify the health check manager of this update.
+  */
+class NotifyHealthCheckManagerStepImpl @Inject() (
+    healthCheckManager: HealthCheckManager) extends TaskStatusUpdateStep {
+  override def name: String = "notifyHealthCheckManager"
+
+  override def processUpdate(
+    timestamp: Timestamp, appId: PathId, maybeTask: Option[MarathonTask], status: TaskStatus): Future[_] = {
+    // forward health changes to the health check manager
+    for (marathonTask <- maybeTask)
+      healthCheckManager.update(status, Timestamp(marathonTask.getVersion))
+
+    Future.successful(())
+  }
+}

--- a/src/main/scala/mesosphere/marathon/core/task/tracker/impl/steps/NotifyLaunchQueueStepImpl.scala
+++ b/src/main/scala/mesosphere/marathon/core/task/tracker/impl/steps/NotifyLaunchQueueStepImpl.scala
@@ -1,0 +1,27 @@
+package mesosphere.marathon.core.task.tracker.impl.steps
+
+import javax.inject.Inject
+
+import mesosphere.marathon.Protos.MarathonTask
+import mesosphere.marathon.core.launchqueue.LaunchQueue
+import mesosphere.marathon.core.task.bus.MarathonTaskStatus
+import mesosphere.marathon.core.task.bus.TaskStatusObservables.TaskStatusUpdate
+import mesosphere.marathon.core.task.tracker.TaskStatusUpdateStep
+import mesosphere.marathon.state.{ PathId, Timestamp }
+import org.apache.mesos.Protos.TaskStatus
+
+import scala.concurrent.Future
+
+/**
+  * Notify the launch queue of this update.
+  */
+class NotifyLaunchQueueStepImpl @Inject() (launchQueue: LaunchQueue) extends TaskStatusUpdateStep {
+  override def name: String = "notifyLaunchQueue"
+
+  override def processUpdate(
+    timestamp: Timestamp, appId: PathId, maybeTask: Option[MarathonTask], status: TaskStatus): Future[_] = {
+    val taskId = status.getTaskId
+    val update = TaskStatusUpdate(timestamp, taskId, MarathonTaskStatus(status))
+    launchQueue.notifyOfTaskUpdate(update)
+  }
+}

--- a/src/main/scala/mesosphere/marathon/core/task/tracker/impl/steps/NotifyRateLimiterStepImpl.scala
+++ b/src/main/scala/mesosphere/marathon/core/task/tracker/impl/steps/NotifyRateLimiterStepImpl.scala
@@ -1,0 +1,53 @@
+package mesosphere.marathon.core.task.tracker.impl.steps
+
+import com.google.inject.Inject
+import mesosphere.marathon.Protos.MarathonTask
+import mesosphere.marathon.core.launchqueue.LaunchQueue
+import mesosphere.marathon.core.task.tracker.TaskStatusUpdateStep
+import mesosphere.marathon.state.{ AppRepository, PathId, Timestamp }
+import org.apache.mesos.Protos.{ TaskState, TaskStatus }
+import org.slf4j.LoggerFactory
+
+import scala.concurrent.Future
+
+class NotifyRateLimiterStepImpl @Inject() (
+    launchQueue: LaunchQueue,
+    appRepository: AppRepository) extends TaskStatusUpdateStep {
+  private[this] val log = LoggerFactory.getLogger(getClass)
+
+  override def name: String = "notifyRateLimiter"
+
+  override def processUpdate(
+    timestamp: Timestamp, appId: PathId, maybeTask: Option[MarathonTask], status: TaskStatus): Future[_] = {
+    import org.apache.mesos.Protos.TaskState._
+
+    status.getState match {
+      case TASK_ERROR | TASK_FAILED | TASK_FINISHED | TASK_KILLED | TASK_LOST =>
+        notifyRateLimiter(appId, status, maybeTask)
+
+      case _ =>
+        Future.successful(())
+    }
+
+  }
+
+  private[this] def notifyRateLimiter(
+    appId: PathId,
+    status: TaskStatus,
+    maybeTask: Option[MarathonTask]): Future[_] = {
+    maybeTask match {
+      case Some(task) if status.getState != TaskState.TASK_KILLED =>
+        import scala.concurrent.ExecutionContext.Implicits.global
+        appRepository.app(appId, Timestamp(task.getVersion)).map { maybeApp =>
+          // It would be nice if we could make sure that the delay gets send
+          // to the AppTaskLauncherActor before we continue but that would require quite some work.
+          //
+          // In production, it should not matter if we do not delay some task launches so much.
+          maybeApp.foreach(launchQueue.addDelay)
+        }
+      case _ =>
+        Future.successful(())
+    }
+  }
+
+}

--- a/src/main/scala/mesosphere/marathon/core/task/tracker/impl/steps/PostToEventStreamStepImpl.scala
+++ b/src/main/scala/mesosphere/marathon/core/task/tracker/impl/steps/PostToEventStreamStepImpl.scala
@@ -1,0 +1,72 @@
+package mesosphere.marathon.core.task.tracker.impl.steps
+
+import javax.inject.Named
+
+import akka.event.EventStream
+import com.google.inject.Inject
+import mesosphere.marathon.Protos.MarathonTask
+import mesosphere.marathon.core.task.tracker.TaskStatusUpdateStep
+import mesosphere.marathon.event.{ EventModule, MesosStatusUpdateEvent }
+import mesosphere.marathon.state.{ PathId, Timestamp }
+import org.apache.mesos.Protos.TaskState.{
+  TASK_ERROR,
+  TASK_FAILED,
+  TASK_FINISHED,
+  TASK_KILLED,
+  TASK_LOST,
+  TASK_RUNNING
+}
+import org.apache.mesos.Protos.TaskStatus
+import org.slf4j.LoggerFactory
+
+import scala.concurrent.Future
+
+/**
+  * Post this update to the internal event stream.
+  */
+class PostToEventStreamStepImpl @Inject() (
+    @Named(EventModule.busName) eventBus: EventStream) extends TaskStatusUpdateStep {
+  private[this] val log = LoggerFactory.getLogger(getClass)
+
+  override def name: String = "postTaskStatusEvent"
+
+  override def processUpdate(
+    timestamp: Timestamp, appId: PathId, maybeTask: Option[MarathonTask], status: TaskStatus): Future[_] = {
+    val taskId = status.getTaskId
+
+    status.getState match {
+      case TASK_ERROR | TASK_FAILED | TASK_FINISHED | TASK_KILLED | TASK_LOST =>
+        // Remove from our internal list
+        maybeTask match {
+          case Some(task) => postEvent(appId, status, task)
+          case None       => log.warn(s"Task not found. Do not post event for '{}'", taskId.getValue)
+        }
+
+      case TASK_RUNNING if !maybeTask.exists(_.hasStartedAt) => // staged, not running
+        maybeTask.foreach { task => postEvent(appId, status, task) }
+
+      case _ =>
+      // ignore
+    }
+
+    Future.successful(())
+  }
+
+  private[this] def postEvent(appId: PathId, status: TaskStatus, task: MarathonTask): Unit = {
+    log.info("Sending event notification for task [{}]: {}", Array[Object](task.getId, status.getState): _*)
+    import scala.collection.JavaConverters._
+    eventBus.publish(
+      MesosStatusUpdateEvent(
+        status.getSlaveId.getValue,
+        status.getTaskId.getValue,
+        status.getState.name,
+        if (status.hasMessage) status.getMessage else "",
+        appId,
+        task.getHost,
+        task.getPortsList.asScala,
+        task.getVersion
+      )
+    )
+  }
+
+}

--- a/src/main/scala/mesosphere/marathon/core/task/tracker/impl/steps/ScaleAppUpdateStepImpl.scala
+++ b/src/main/scala/mesosphere/marathon/core/task/tracker/impl/steps/ScaleAppUpdateStepImpl.scala
@@ -1,0 +1,43 @@
+package mesosphere.marathon.core.task.tracker.impl.steps
+
+import javax.inject.Named
+
+import akka.actor.ActorRef
+import com.google.inject.Inject
+import mesosphere.marathon.MarathonSchedulerActor.ScaleApp
+import mesosphere.marathon.Protos.MarathonTask
+import mesosphere.marathon.core.task.tracker.TaskStatusUpdateStep
+import mesosphere.marathon.state.{ AppDefinition, PathId, Timestamp }
+import org.apache.mesos.Protos.TaskStatus
+import org.slf4j.LoggerFactory
+
+import scala.concurrent.Future
+
+/**
+  * Trigger rescale of affected app if a task died.
+  */
+class ScaleAppUpdateStepImpl @Inject() (
+    @Named("schedulerActor") schedulerActor: ActorRef) extends TaskStatusUpdateStep {
+  private[this] val log = LoggerFactory.getLogger(getClass)
+
+  override def name: String = "scaleApp"
+
+  override def processUpdate(
+    timestamp: Timestamp, appId: PathId, maybeTask: Option[MarathonTask], status: TaskStatus): Future[_] = {
+    val taskId = status.getTaskId
+
+    import org.apache.mesos.Protos.TaskState._
+
+    status.getState match {
+      case TASK_ERROR | TASK_FAILED | TASK_FINISHED | TASK_KILLED | TASK_LOST =>
+        // Remove from our internal list
+        log.info(s"initiating a scale check for app [$appId] after task [${taskId.getValue}}] terminated")
+        schedulerActor ! ScaleApp(appId)
+
+      case _ =>
+      // ignore
+    }
+
+    Future.successful(())
+  }
+}

--- a/src/main/scala/mesosphere/marathon/core/task/tracker/impl/steps/TaskStatusEmitterPublishStepImpl.scala
+++ b/src/main/scala/mesosphere/marathon/core/task/tracker/impl/steps/TaskStatusEmitterPublishStepImpl.scala
@@ -1,0 +1,26 @@
+package mesosphere.marathon.core.task.tracker.impl.steps
+
+import javax.inject.Inject
+
+import mesosphere.marathon.Protos.MarathonTask
+import mesosphere.marathon.core.task.bus.TaskStatusObservables.TaskStatusUpdate
+import mesosphere.marathon.core.task.bus.{ MarathonTaskStatus, TaskStatusEmitter }
+import mesosphere.marathon.core.task.tracker.TaskStatusUpdateStep
+import mesosphere.marathon.state.{ PathId, Timestamp }
+import org.apache.mesos.Protos.TaskStatus
+
+import scala.concurrent.Future
+
+/**
+  * Forward the update to the taskStatusEmitter.
+  */
+class TaskStatusEmitterPublishStepImpl @Inject() (taskStatusEmitter: TaskStatusEmitter) extends TaskStatusUpdateStep {
+  override def name: String = "emitUpdate"
+
+  override def processUpdate(
+    timestamp: Timestamp, appId: PathId, maybeTask: Option[MarathonTask], status: TaskStatus): Future[_] = {
+    val taskId = status.getTaskId
+    taskStatusEmitter.publish(TaskStatusUpdate(timestamp, taskId, MarathonTaskStatus(status)))
+    Future.successful(())
+  }
+}

--- a/src/main/scala/mesosphere/marathon/core/task/tracker/impl/steps/UpdateTaskTrackerStepImpl.scala
+++ b/src/main/scala/mesosphere/marathon/core/task/tracker/impl/steps/UpdateTaskTrackerStepImpl.scala
@@ -1,0 +1,65 @@
+package mesosphere.marathon.core.task.tracker.impl.steps
+
+import com.google.inject.Inject
+import mesosphere.marathon.MarathonSchedulerDriverHolder
+import mesosphere.marathon.Protos.MarathonTask
+import mesosphere.marathon.core.task.tracker.TaskStatusUpdateStep
+import mesosphere.marathon.state.{ AppDefinition, PathId, Timestamp }
+import mesosphere.marathon.tasks.TaskTracker
+import org.apache.mesos.Protos.TaskStatus
+import org.slf4j.LoggerFactory
+
+import scala.concurrent.Future
+import scala.util.control.NonFatal
+
+/**
+  * Update task tracker corresponding to the event.
+  */
+class UpdateTaskTrackerStepImpl @Inject() (
+    taskTracker: TaskTracker,
+    marathonSchedulerDriverHolder: MarathonSchedulerDriverHolder) extends TaskStatusUpdateStep {
+
+  private[this] val log = LoggerFactory.getLogger(getClass)
+
+  def name: String = "updateTaskTracker"
+
+  //TODO: fix style issue and enable this scalastyle check
+  //scalastyle:off cyclomatic.complexity method.length
+  def processUpdate(
+    timestamp: Timestamp, appId: PathId, maybeTask: Option[MarathonTask], status: TaskStatus): Future[_] = {
+    val taskId = status.getTaskId
+
+    import org.apache.mesos.Protos.TaskState._
+
+    import scala.concurrent.ExecutionContext.Implicits.global
+    status.getState match {
+      case TASK_ERROR | TASK_FAILED | TASK_FINISHED | TASK_KILLED | TASK_LOST =>
+        // Remove from our internal list
+        taskTracker.terminated(appId, taskId.getValue)
+
+      case TASK_RUNNING if !maybeTask.exists(_.hasStartedAt) => // staged, not running
+        taskTracker.running(appId, status)
+          .recover {
+            case NonFatal(t) =>
+              log.warn(s"Task could not be saved: '${taskId.getValue}'", t)
+              driverOpt.foreach(_.killTask(status.getTaskId))
+          }
+
+      case TASK_STAGING if !taskTracker.contains(appId) =>
+        log.warn(s"Received status update for unknown app $appId, killing task ${status.getTaskId}")
+        driverOpt.foreach(_.killTask(status.getTaskId))
+        Future.successful(())
+
+      case _ =>
+        taskTracker.statusUpdate(appId, status).map {
+          case None =>
+            log.warn(s"Killing task ${status.getTaskId}")
+            driverOpt.foreach(_.killTask(status.getTaskId))
+          case _ =>
+        }
+    }
+  }
+
+  private[this] def driverOpt = marathonSchedulerDriverHolder.driver
+
+}

--- a/src/test/scala/mesosphere/marathon/core/launchqueue/impl/AppTaskLauncherActorTest.scala
+++ b/src/test/scala/mesosphere/marathon/core/launchqueue/impl/AppTaskLauncherActorTest.scala
@@ -201,7 +201,7 @@ class AppTaskLauncherActorTest extends MarathonSpec with GivenWhenThen {
     val props = Props(
       new AppTaskLauncherActor(
         launchQueueConfig,
-        offerMatcherManager, clock, taskFactory, taskStatusObservable,
+        offerMatcherManager, clock, taskFactory,
         maybeOfferReviver = None,
         taskTracker, rateLimiterActor.ref,
         app, tasksToLaunch = 1
@@ -276,9 +276,10 @@ class AppTaskLauncherActorTest extends MarathonSpec with GivenWhenThen {
       Await.result(launcherRef ? AppTaskLauncherActor.GetCount, 3.seconds).asInstanceOf[QueuedTaskCount]
 
       // task status update
-      appStatusObservable.onNext(update.withTaskId(marathonTask.getId).wrapped)
-
-      val counts = Await.result(launcherRef ? AppTaskLauncherActor.GetCount, 3.seconds).asInstanceOf[QueuedTaskCount]
+      val counts = Await.result(
+        launcherRef ? update.withTaskId(marathonTask.getId).wrapped,
+        3.seconds
+      ).asInstanceOf[QueuedTaskCount]
 
       assert(counts.tasksLaunchedOrRunning == 0)
 
@@ -315,10 +316,7 @@ class AppTaskLauncherActorTest extends MarathonSpec with GivenWhenThen {
       Await.result(launcherRef ? AppTaskLauncherActor.GetCount, 3.seconds).asInstanceOf[QueuedTaskCount]
 
       When("we get a status update about a terminated task")
-      appStatusObservable.onNext(update.withTaskId(marathonTask.getId).wrapped)
-
-      And("this update has been fully processed")
-      Await.result(launcherRef ? AppTaskLauncherActor.GetCount, 3.seconds).asInstanceOf[QueuedTaskCount]
+      Await.result(launcherRef ? update.withTaskId(marathonTask.getId).wrapped, 3.seconds).asInstanceOf[QueuedTaskCount]
 
       Then("reviveOffers has been called")
       Mockito.verify(offerReviver).reviveOffers()
@@ -344,9 +342,10 @@ class AppTaskLauncherActorTest extends MarathonSpec with GivenWhenThen {
       Await.result(launcherRef ? AppTaskLauncherActor.GetCount, 3.seconds).asInstanceOf[QueuedTaskCount]
 
       // task status update
-      appStatusObservable.onNext(update.withTaskId(marathonTask.getId).wrapped)
-
-      val counts = Await.result(launcherRef ? AppTaskLauncherActor.GetCount, 3.seconds).asInstanceOf[QueuedTaskCount]
+      val counts = Await.result(
+        launcherRef ? update.withTaskId(marathonTask.getId).wrapped,
+        3.seconds
+      ).asInstanceOf[QueuedTaskCount]
 
       assert(counts.tasksLaunchedOrRunning == 1)
 
@@ -368,8 +367,6 @@ class AppTaskLauncherActorTest extends MarathonSpec with GivenWhenThen {
   private[this] var offerMatcherManager: OfferMatcherManager = _
   private[this] var clock: ConstantClock = _
   private[this] var taskFactory: TaskFactory = _
-  private[this] var appStatusObservable: Subject[TaskStatusUpdate] = _
-  private[this] var taskStatusObservable: TaskStatusObservables = _
   private[this] var taskTracker: TaskTracker = _
   private[this] var offerReviver: OfferReviver = _
   private[this] var rateLimiterActor: TestProbe = _
@@ -377,7 +374,7 @@ class AppTaskLauncherActorTest extends MarathonSpec with GivenWhenThen {
   private[this] def createLauncherRef(instances: Int, appToLaunch: AppDefinition = app): ActorRef = {
     val props = AppTaskLauncherActor.props(
       launchQueueConfig,
-      offerMatcherManager, clock, taskFactory, taskStatusObservable,
+      offerMatcherManager, clock, taskFactory,
       maybeOfferReviver = Some(offerReviver),
       taskTracker, rateLimiterActor.ref) _
     actorSystem.actorOf(
@@ -393,9 +390,6 @@ class AppTaskLauncherActorTest extends MarathonSpec with GivenWhenThen {
     launchQueueConfig.afterInit()
     clock = ConstantClock()
     taskFactory = mock[TaskFactory]
-    appStatusObservable = PublishSubject[TaskStatusUpdate]()
-    taskStatusObservable = mock[TaskStatusObservables]
-    Mockito.when(taskStatusObservable.forAppId(app.id)).thenReturn(appStatusObservable)
     taskTracker = mock[TaskTracker]
     offerReviver = mock[OfferReviver]
     rateLimiterActor = TestProbe()
@@ -406,8 +400,6 @@ class AppTaskLauncherActorTest extends MarathonSpec with GivenWhenThen {
     // Mockito.verifyNoMoreInteractions(offerMatcherManager)
     Mockito.verifyNoMoreInteractions(taskFactory)
     Mockito.verifyNoMoreInteractions(taskTracker)
-    Mockito.verify(taskStatusObservable).forAppId(app.id)
-    Mockito.verifyNoMoreInteractions(taskStatusObservable)
 
     actorSystem.shutdown()
     actorSystem.awaitTermination()

--- a/src/test/scala/mesosphere/marathon/core/launchqueue/impl/LaunchQueueModuleTest.scala
+++ b/src/test/scala/mesosphere/marathon/core/launchqueue/impl/LaunchQueueModuleTest.scala
@@ -196,7 +196,6 @@ class LaunchQueueModuleTest extends MarathonSpec with BeforeAndAfter with GivenW
       AlwaysElectedLeadershipModule(shutdownHooks),
       clock,
       subOfferMatcherManager = offerMatcherManager,
-      taskStatusObservables = taskBusModule.taskStatusObservables,
       maybeOfferReviver = None,
       appRepository,
       taskTracker,

--- a/src/test/scala/mesosphere/marathon/core/task/tracker/impl/TaskStatusUpdateProcessorImplTest.scala
+++ b/src/test/scala/mesosphere/marathon/core/task/tracker/impl/TaskStatusUpdateProcessorImplTest.scala
@@ -2,40 +2,41 @@ package mesosphere.marathon.core.task.tracker.impl
 
 import akka.actor.ActorSystem
 import akka.event.EventStream
-import akka.testkit.{ TestActorRef, TestProbe }
+import akka.testkit.TestProbe
+import com.codahale.metrics.MetricRegistry
 import mesosphere.marathon.MarathonSchedulerActor.ScaleApp
 import mesosphere.marathon.Protos.MarathonTask
+import mesosphere.marathon.core.base.ConstantClock
 import mesosphere.marathon.core.launchqueue.LaunchQueue
-import mesosphere.marathon.core.leadership.PreparationMessages
-import mesosphere.marathon.core.task.bus.{ TaskStatusEmitter, TaskStatusUpdateTestHelper, TaskStatusObservables }
-import mesosphere.marathon.core.task.bus.TaskStatusObservables.TaskStatusUpdate
+import mesosphere.marathon.core.task.bus.{TaskStatusEmitter, TaskStatusUpdateTestHelper}
 import mesosphere.marathon.core.task.tracker.TaskStatusUpdateProcessor
+import mesosphere.marathon.core.task.tracker.impl.steps.{AcknowledgeTaskUpdateStepImpl, NotifyHealthCheckManagerStepImpl, NotifyRateLimiterStepImpl, PostToEventStreamStepImpl, ScaleAppUpdateStepImpl, TaskStatusEmitterPublishStepImpl, UpdateTaskTrackerStepImpl}
 import mesosphere.marathon.event.MesosStatusUpdateEvent
 import mesosphere.marathon.health.HealthCheckManager
-import mesosphere.marathon.state.{ AppDefinition, AppRepository, Timestamp, PathId }
-import mesosphere.marathon.tasks.{ TaskIdUtil, TaskTracker }
-import mesosphere.marathon.{ MarathonTestHelper, MarathonSchedulerDriverHolder, MarathonSpec }
-import org.apache.mesos.Protos.{ TaskState, TaskStatus }
+import mesosphere.marathon.metrics.Metrics
+import mesosphere.marathon.state.{AppDefinition, AppRepository, PathId, Timestamp}
+import mesosphere.marathon.tasks.{TaskIdUtil, TaskTracker}
+import mesosphere.marathon.{MarathonSchedulerDriverHolder, MarathonSpec, MarathonTestHelper}
+import org.apache.mesos.Protos.TaskState
 import org.apache.mesos.SchedulerDriver
-import org.mockito.internal.matchers.CapturingMatcher
-import org.mockito.{ ArgumentCaptor, Matchers, Mockito }
-import rx.lang.scala.Subject
-import rx.lang.scala.subjects.PublishSubject
+import org.mockito.{ArgumentCaptor, Mockito}
 
-import scala.concurrent.{ Await, Future }
 import scala.concurrent.duration._
+import scala.concurrent.{Await, Future}
 
 class TaskStatusUpdateProcessorImplTest extends MarathonSpec {
 
   for (
-    update <- Seq(
+    origUpdate <- Seq(
       TaskStatusUpdateTestHelper.finished,
       TaskStatusUpdateTestHelper.lost,
       TaskStatusUpdateTestHelper.killed,
       TaskStatusUpdateTestHelper.error
-    ).map(_.withAppId(appId.toString))
+    )
   ) {
-    test(s"Remove terminated task (${update.wrapped.status.getClass.getSimpleName})") {
+    test(s"Remove terminated task (${origUpdate.wrapped.status.getClass.getSimpleName})") {
+      val status = origUpdate.wrapped.status.mesosStatus.get.toBuilder.setTaskId(TaskIdUtil.newTaskId(appId)).build()
+      val update = origUpdate.withTaskId(status.getTaskId)
 
       Mockito.when(taskTracker.fetchTask(appId, update.wrapped.taskId.getValue))
         .thenReturn(Some(marathonTask))
@@ -43,10 +44,9 @@ class TaskStatusUpdateProcessorImplTest extends MarathonSpec {
         .thenReturn(Future.successful(Some(marathonTask)))
       Mockito.when(appRepository.app(appId, version)).thenReturn(Future.successful(Some(app)))
 
-      Await.result(updateProcessor.publish(update.wrapped), 3.seconds)
+      Await.result(updateProcessor.publish(status), 3.seconds)
 
       Mockito.verify(taskTracker).fetchTask(appId, update.wrapped.taskId.getValue)
-      val status: TaskStatus = update.wrapped.status.mesosStatus.get
       Mockito.verify(healthCheckManager).update(status, version)
       Mockito.verify(taskTracker).terminated(appId, update.wrapped.taskId.getValue)
       schedulerActor.expectMsg(ScaleApp(appId))
@@ -72,6 +72,7 @@ class TaskStatusUpdateProcessorImplTest extends MarathonSpec {
     MarathonTask.newBuilder().setId(task.getTaskId.getValue).setVersion(version.toString).build()
 
   private[this] implicit var actorSystem: ActorSystem = _
+  private[this] var clock: ConstantClock = _
   private[this] var taskStatusEmitter: TaskStatusEmitter = _
   private[this] var appRepository: AppRepository = _
   private[this] var launchQueue: LaunchQueue = _
@@ -86,6 +87,7 @@ class TaskStatusUpdateProcessorImplTest extends MarathonSpec {
 
   before {
     actorSystem = ActorSystem()
+    clock = ConstantClock()
     taskStatusEmitter = mock[TaskStatusEmitter]
     appRepository = mock[AppRepository]
     launchQueue = mock[LaunchQueue]
@@ -98,16 +100,37 @@ class TaskStatusUpdateProcessorImplTest extends MarathonSpec {
     marathonSchedulerDriverHolder = new MarathonSchedulerDriverHolder
     marathonSchedulerDriverHolder.driver = Some(schedulerDriver)
 
-    updateProcessor = new TaskStatusUpdateProcessorImpl(
-      taskStatusEmitter,
-      appRepository,
-      launchQueue,
-      eventBus,
-      schedulerActor.ref,
-      taskIdUtil,
-      healthCheckManager,
+    val notifyHealthCheckManager = new NotifyHealthCheckManagerStepImpl(healthCheckManager)
+
+    val notifyRateLimiter = new NotifyRateLimiterStepImpl(launchQueue, appRepository)
+
+    val acknowledgeStep = new AcknowledgeTaskUpdateStepImpl(marathonSchedulerDriverHolder)
+
+    val updateTaskTrackerStep = new UpdateTaskTrackerStepImpl(
       taskTracker,
       marathonSchedulerDriverHolder
+    )
+
+    val postToEventStream = new PostToEventStreamStepImpl(eventBus)
+
+    val emitUpdate = new TaskStatusEmitterPublishStepImpl(taskStatusEmitter)
+
+    val scaleApp = new ScaleAppUpdateStepImpl(schedulerActor.ref)
+
+    updateProcessor = new TaskStatusUpdateProcessorImpl(
+      new Metrics(new MetricRegistry),
+      clock,
+      taskIdUtil,
+      taskTracker,
+      Seq(
+        notifyHealthCheckManager,
+        notifyRateLimiter,
+        updateTaskTrackerStep,
+        emitUpdate,
+        postToEventStream,
+        scaleApp,
+        acknowledgeStep
+      )
     )
   }
 

